### DISCUSSION TITLE
C++

### DISCUSSION THREAD

**Discussion by darkAlert** [kornia/kornia#1750] [General]
Hi, 
Are there ways to use kornia in C++ on Jetson?

**Comment by edgarriba** [comment]
@darkAlert one option is to create a module that embeds a function (or group) and just torchscript. Ideally, at some point we would like to provide 1:1 mapping interface but low on resources now. Do you have any specific function request ? we could try to discuss from where to start. I personally start to need the same thing at work :)

**Reply by darkAlert** [reply]
thank you for the tip with torchscript, will try it

My colleague and I are trying to stitch images from multiple cameras into a panorama (using cv::warpPerspective on Jetson TX2 NX). We tried opencv with CPU but it is very slow (2 fps), and opencv with CUDA (20 FPS). But we want at least 30 FPS. 
From my experience on my desktop computer, kornia works faster then cv::cuda. So it would be interesting to test kornia.geometry.transform.warp_perspective on Jetson.

**Reply by edgarriba** [reply]
something I would try first is to benchmark from python, then we can explore to implement warpPerspective in c++

I think it's a good first operator to add. We need to discuss later how to integrate and maintain the c++ codebase. We are starting very a collaboration with the company I work to improve the CI, so I guess it's a good timing. I believe I can spend some time to prototype the warpPerspective function if needed.

**Reply by darkAlert** [reply]
great, thanks for your work! As for me, warpPerspective is a very useful function (and HomographyWarper class as well), I use them very often in different models (but maybe this is because I work with homography :) ) 

BTW, I recently found out that now mode='nearest' works! Previously (~1.5 year ago) I tried to use it during training, but it seems that gradient didn't work correctly and I've been forced to use 'bilinear' mode so far. So it's great that you fixed it!

**Reply by edgarriba** [reply]
awesome ! then let's make it happen. Probably we could target this as a cpp extension ? and expose both as  python and c++
> https://pytorch.org/tutorials/advanced/cpp_extension.html

**Reply by edgarriba** [reply]
one more thought in terms of maintenance would be how we setup the ci and package. I'm setting up some sponsorships on that end because the complexity to package c++/cuda is not as simple as python. Any thoughts/help here would be very welcomed

**Reply by edgarriba** [reply]
one more option in order to re use kornia python code would be to torchscript the function and setup a header only c++ lib with dynamically loaded libraries. Not sure though about the performance here -- but definitely will help to easily hack the code from python side. Modules can be implemented in terms of pimpl pattern.
